### PR TITLE
haku: adaptive surface_pressure loss-weight decay (1.0→0.3)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Adaptive per-channel loss weighting** — start with the current static weights (`sp=1.0, ws_x=1.0, ws_y=2.0, ws_z=2.0, vp=1.0`), and over the course of training, monotonically decay the surface_pressure loss weight from `1.0 → 0.3` linearly across the budget. The motivation:

- Surface pressure converges fastest on yi (~5.5% at convergence vs 8.5–10% for wall_shear channels). Once it reaches a low residual, additional gradient signal on it is **wasted capacity** that could be flowing into harder channels (wall_shear_y/z, volume_pressure).
- The current static weighting (`sp=1, ws_y=2, ws_z=2`) bakes in an a-priori judgment about channel difficulty. This experiment tests a **time-varying** weighting that lets the optimization budget reflow toward the channels that haven't converged yet.
- This is **gradient capacity reallocation** by direct loss-weight scheduling — orthogonal to PR #529 (stream-normal per-point weighting, which is geometric) and PR #518 (asinh residual transform, which is robust-loss). It does not change the loss form, just its time-varying scalar weight.

This is a closer, more controllable cousin of GradNorm (Chen et al. 2017) without the auxiliary network — we use a fixed schedule rather than learning the weights, since the convergence ordering on yi is already known empirically.

### Code change

Add a new flag `--surface-pressure-loss-weight-decay <final_value>` (default `0.0` = OFF, i.e. constant weight). When set to a value `w_final` < 1.0:

1. Locate the loss-weighting block in `target/train.py` where `surface_pressure_loss_weight` (or whatever the variable is named — likely `sp_w` or similar; trace from the existing static config field) is multiplied with the surface_pressure residual.

2. Replace the static weight with a time-varying linear schedule:
```python
if config.surface_pressure_loss_weight_decay > 0.0:
    progress = global_step / max(1, total_steps)  # 0.0 at start, 1.0 at end
    progress = min(1.0, progress)
    w_init = 1.0  # initial weight
    w_final = config.surface_pressure_loss_weight_decay
    sp_w_now = w_init + progress * (w_final - w_init)  # linear interpolation
else:
    sp_w_now = config.surface_pressure_loss_weight  # static existing value
```

3. Log `train/sp_loss_weight_now` every `gradient_log_every` steps so the schedule is visible in W&B.

4. **Do not modify** the wall_shear or volume_pressure weight code paths. This is a one-channel intervention.

### Training run — single arm DDP-4

```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku \
  --wandb-group haku-r32-sp-loss-decay \
  --wandb-name haku-r32-sp-decay-1.0-to-0.3 \
  --optimizer lion \
  --lr 1e-4 --weight-decay 1e-4 \
  --no-compile-model --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1 --clip-grad-norm 0.5 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --epochs 50 \
  --surface-pressure-loss-weight-decay 0.3
```

Single arm only — this is a hypothesis-test, not a sweep. If it wins, we sweep `0.5 / 0.3 / 0.1` afterwards.

### Reporting

| Metric | yi baseline (PR #309 control) | This run | Δ |
|---|---:|---:|---:|
| val_abupt | 9.039% | | |
| surface_pressure (val) | ~5.5% | | |
| wall_shear (val) | ~10.5% | | |
| wall_shear_y (val) | ~10.5% | | |
| wall_shear_z (val) | ~11.8% | | |
| volume_pressure (val) | ~13.5% | | |
| test_abupt | ~10.2% | | |

Plus: best `train/sp_loss_weight_now` trajectory across training (W&B chart), best epoch, W&B run ID.

## What success looks like

- val_abupt < 9.039% bar → MERGE
- surface_pressure regresses ≤ +0.5pp (acceptable trade) AND wall_shear or volume_pressure improves ≥ 0.5pp → mechanism confirmed even if val_abupt is null; follow-up sweeps justified
- All channels regress → adaptive weighting is hurting; close
- val_abupt improves but surface_pressure regresses > 1.0pp → still merge (val_abupt is the primary metric), but flag the surface trade-off

## Baseline

- **Active merge bar:** `val_primary/abupt_axis_mean_rel_l2_pct < 9.039%` (PR #309 thorfinn grad-clip=0.5 on yi).
- **Aspirational tay number for context:** 7.546% (PR #311) and 7.1792% (PR #489) are tay-branch results, NOT on yi. They are not the merge bar for this experiment.

## Notes

- This is a **time-varying weighting**, not a static one. The whole point of the experiment is the schedule. Confirm the W&B-logged `sp_loss_weight_now` is monotonically decreasing across training.
- `--epochs 50` keeps the schedule denominator at the canonical config; actual training will hit timeout at ~3 epochs ≈ 16k steps, so the surface-pressure weight at termination will be `1.0 + (16k/272k) × (0.3 - 1.0) ≈ 0.96`. This means the schedule is barely engaged — that's a problem.
- **Better: use `--epochs 3` so the schedule fires properly.** With `--epochs 3` and ~16k total steps, the weight transitions linearly from 1.0 to 0.3 over the actual run — the schedule fully engages. Use `--epochs 3` in your launch.

(Replace `--epochs 50` in the command above with `--epochs 3` for actual launch.)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Haku — strong hypothesis, well-motivated by the known convergence ordering on yi. The PR body is correct in every key detail. Approved to launch.

**Key points before you start:**

1. **Use `--epochs 3` NOT `--epochs 50`** — as you correctly noted in the PR body, with `--epochs 50` only ~6% of the schedule fires (sp_loss_weight barely moves from 1.0 to ~0.96). With `--epochs 3` the weight transitions fully from 1.0 → 0.3 over the actual run. The schedule only works if you use `--epochs 3`.

2. **Verify the W&B log of `train/sp_loss_weight_now`** — confirm it starts at 1.0 and monotonically decreases toward ~0.3 over the run. A flat line at 1.0 means the schedule is not firing.

3. **Single arm only** — correct, no sweep until we see if this direction works. Post results as a PR comment with the full table from the PR body template.

**Corrected launch command:**
```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku \
  --wandb-group haku-r32-sp-loss-decay \
  --wandb-name haku-r32-sp-decay-1.0-to-0.3 \
  --optimizer lion \
  --lr 1e-4 --weight-decay 1e-4 \
  --no-compile-model --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1 --clip-grad-norm 0.5 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --epochs 3 \
  --surface-pressure-loss-weight-decay 0.3
```

**Merge bar: val_abupt < 9.039%** (PR #309 yi). tay-branch numbers (7.546% / 7.1792%) are NOT the yi bar.

Launch now and confirm with W&B run ID.

**Comment by morganmcg1** [comment]
ADVISOR: Run `zcb17qa5` confirmed in flight at step ~1,400. Watch list:

1. **`train/sp_loss_weight_now` schedule must be visible in W&B**: verify it starts at 1.0 and monotonically decreases toward 0.3 over the actual training duration (~12,750 steps target). If you see a flat line at 1.0, the schedule isn't firing — halt and report.

2. **First val_primary (epoch 1) per-channel breakdown**: this is where we expect to see the trade — surface_pressure rel_l2 should rise slightly relative to a no-decay baseline (because we're down-weighting it), while wall_shear / volume_pressure should improve. The hypothesis succeeds only if the **net val_primary** drops below 9.039%.

3. **Final report should include**:
   - All per-epoch val_primary
   - Per-channel breakdown (surface_pressure, wall_shear, wall_shear_y, wall_shear_z, volume_pressure)
   - Best-EMA-checkpoint test_primary (NOT terminal)
   - The actual sp_loss_weight at the final step (should be ~0.3)
   - Step count + wall-clock at exit

Send back when run completes. Marking WIP.

**Comment by morganmcg1** [comment]
STUDENT haku:

## Results

**Verdict: FAILS merge bar** — `val_primary/abupt_axis_mean_rel_l2_pct = 10.674%` vs `9.039%` baseline (+1.64pp). All four `abupt`-axis channels (sp, ws_x, ws_y, ws_z) regressed; `volume_pressure` is the only channel that improved.

The schedule fired correctly (W&B `train/sp_loss_weight_now`: 1.0 → 0.471 over 12,343 steps), but the run hit the 270-min wall-clock timeout mid-epoch 3 — it would have reached the planned `w_final = 0.3` at step 16,326 (≈3 full epochs).

### Validation per channel (epoch 3, best checkpoint)

| Metric | yi baseline (PR body) | This run | Δ |
|---|---:|---:|---:|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **9.039** | **10.674** | **+1.64** |
| `val/surface_pressure_rel_l2_pct` | ~5.5 | 7.349 | +1.85 |
| `val/wall_shear_rel_l2_pct` (vec) | ~10.5 | 11.920 | +1.42 |
| `val/wall_shear_x_rel_l2_pct` | — | 10.300 | — |
| `val/wall_shear_y_rel_l2_pct` | ~10.5 | 14.440 | +3.94 |
| `val/wall_shear_z_rel_l2_pct` | ~11.8 | 14.623 | +2.82 |
| `val/volume_pressure_rel_l2_pct` | ~13.5 | **6.661** | **−6.84** |
| `val/surface_pressure_mae` | — | 0.02047 | — |
| `val/wall_shear_mae` | — | 0.10839 | — |
| `val/volume_pressure_mae` | — | 12.4091 | — |

(`test_primary/*` was NOT reached — timeout forced exit before the post-train test phase.)

### Per-epoch trajectory

| Epoch | Step | val_abupt | sp | ws | ws_y | ws_z | vp |
|---:|---:|---:|---:|---:|---:|---:|---:|
| 1 | 5,442 | 22.312 | 15.917 | 24.467 | 29.040 | 30.255 | 14.997 |
| 2 | 10,885 | 11.671 | 8.006 | 12.990 | 15.640 | 15.993 | 7.466 |
| 3 (timeout) | 12,343 | 10.674 | 7.349 | 11.920 | 14.440 | 14.623 | 6.661 |

### Schedule sanity-check (sp_loss_weight_now)

| Step | weight |
|---:|---:|
| 1 | 1.0000 |
| 2,469 | 0.8942 |
| 6,170 | 0.7355 |
| 9,873 | 0.5767 |
| 12,343 (terminal) | 0.4708 |

Per-1k-step slope: −0.0429 — clean monotonic linear decay, exactly the linear interpolation the PR specified. Schedule mechanics are correct.

### Run details

- **W&B run:** `zcb17qa5` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/zcb17qa5))
- **Wall-clock:** 275.6 min (terminated by 270-min train-loop timeout at step 12,343)
- **Steps completed:** 12,343 of planned 16,326 (75.6% of schedule)
- **Final sp_loss_weight:** 0.4708 (planned 0.3 at completion)
- **Peak memory:** 67.5 GB / GPU (4× RTX PRO 6000 Blackwell, ~68% memoryAllocated)
- **Effective batch:** 16 (4 per GPU × 4 GPUs DDP)

### Exact command

```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku \
  --wandb-group haku-r32-sp-loss-decay \
  --wandb-name haku-r32-sp-decay-1.0-to-0.3 \
  --optimizer lion \
  --lr 1e-4 --weight-decay 1e-4 \
  --no-compile-model --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1 --clip-grad-norm 0.5 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --epochs 3 \
  --surface-pressure-loss-weight-decay 0.3
```

### What happened

**The hypothesis failed on val_abupt.** The intended mechanism — that down-weighting `surface_pressure` would reflow gradient capacity to `wall_shear_y/z` (the slowest-converging channels) — did not occur. Instead:

- Every `sp`/`ws_*` channel **regressed** (sp +1.85pp, ws +1.42pp, ws_y +3.94pp, ws_z +2.82pp).
- `volume_pressure` **improved dramatically** (−6.84pp vs the 13.5% number quoted in the PR baseline; from 14.997 → 6.661 within this run alone).

So gradient capacity *was* reallocated away from surface_pressure, but it flowed almost entirely into the **volume head**, not into wall_shear. This is a surprising mechanism finding — the surface and volume heads share the backbone but have separate output paths, so down-weighting any *surface* channel apparently lets the optimization budget bias toward volume targets, not toward other surface channels.

Two confounds I want to flag:

1. **Schedule was truncated at 75.6%** — the weight only reached 0.471, not the planned 0.3. The "fully-engaged" schedule is not what got tested. A more aggressive schedule (e.g. faster decay, or `--epochs 2.27` denominator) would let the full 1.0 → 0.3 trajectory fire within the wall-clock budget.
2. **The `~13.5%` `volume_pressure` baseline in the PR body looks stale** — even at epoch 2 of this run (with minimal sp-decay applied) `vp = 7.466%`, far below the 13.5% reference. If a fresh PR #309-style yi control already gets vp ≈ 7%, then most of the apparent vp gain here is not from the decay schedule. I'd recommend the advisor cross-check against a fresh yi control run.

Per the PR's success criteria: this is a clear "all (primary) channels regress" outcome on the abupt-axis metrics, which the PR body says should close the direction. I'm reporting it as such.

### Suggested follow-ups

1. **Cross-check the `volume_pressure` baseline.** If a same-config yi control (no sp-decay) gets vp ≈ 7%, then the dramatic vp improvement here is just the modern yi baseline, not the decay mechanism, and there's nothing to chase. If vp ≈ 13% on the yi control, then the **decay-as-volume-booster** is a real mechanism worth testing on its own (e.g. could combine with stronger wall_shear weights).
2. **Try a *faster* decay schedule that completes within the wall-clock budget.** With `--epochs 3` the schedule denominator gives 16,326 steps but training only runs ≈12,343 steps before timeout. Setting the schedule to terminate at the actual step budget (e.g. via a `--sp-decay-steps 12000` flag, or just lowering `--epochs` to ≈2.3) lets the full 1.0 → 0.3 sweep fire and tests the hypothesis as designed.
3. **The wall_shear regression is the smoking gun against the original mechanism hypothesis.** ws_y went from a ~10.5% baseline to 14.4% here — substantially worse, not better. This argues that a flat down-weighting of surface_pressure does *not* free capacity for hard surface channels. If the goal is to help wall_shear, a more surgical intervention (e.g. directly raising ws_y/ws_z weights, or using stream-normal weighting per PR #529) is more likely to work than indirectly hoping a sp down-weight pushes capacity sideways.
4. **If the advisor wants to revisit, do not sweep `0.5 / 0.3 / 0.1`** as the original PR body proposed — the val_abupt regression is consistent across the whole schedule range that fired (1.0 → 0.47), and a more aggressive endpoint will only make the surface_pressure regression worse without solving the wall_shear problem.

I think this direction is closed. Awaiting advisor review.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #533 because ADVISOR: Closing — clear dead end. W&B confirmed: val_primary/abupt = 10.674% vs 9.039% baseline (+1.635pp). All four abupt-axis surface channels regressed (sp +1.85pp, ws +1.42pp, ws_y +3.94pp, ws_z +2.82pp).

The mechanism hypothesis was falsified: down-weighting surface_pressure did not reflow gradient capacity to wall_shear_y/z. Instead, it flowed almost entirely into the volume head. The volume_pressure gain (6.661%) is real mechanistically but the ~13.5% baseline in the PR body was stale — a fresh yi control likely already sits at ~7%, so the apparent gain is not attributable to this schedule.

The student's per-step slope analysis confirms ws_z is still descending at -0.94pp/1k steps — the model would improve with more epochs, but the direction itself is not sound. Following haku's recommendation: do NOT sweep 0.5/0.3/0.1. The consistent ws_y regression across the whole weight range that fired (1.0→0.47) rules out an endpoint sweep recovering the metric.

The hypothesis is closed. Well-run experiment with good diagnostics..
